src/psset.c

```diff
@@ -0,0 +1,211 @@
+ */
+static edata_t *
+psset_recycle_extract(psset_t *psset, size_t size) {
+	pszind_t ret_ind;
+	edata_t *ret = NULL;
+	pszind_t pind = sz_psz2ind(sz_psz_quantize_ceil(size));
+	for (pszind_t i = (pszind_t)bitmap_ffu(psset->bitmap,
+	    &psset_bitmap_info, (size_t)pind);
+	    i < PSSET_NPSIZES;
+	    i = (pszind_t)bitmap_ffu(psset->bitmap, &psset_bitmap_info,
+		(size_t)i + 1)) {
+		assert(!edata_heap_empty(&psset->pageslabs[i]));
+		edata_t *ps = edata_heap_first(&psset->pageslabs[i]);
+		if (ret == NULL || edata_snad_comp(ps, ret) < 0) {
+			ret = ps;
+			ret_ind = i;
+		}
+	}
+	if (ret == NULL) {
+		return NULL;
+	}
+	edata_heap_remove(&psset->pageslabs[ret_ind], ret);
+	if (edata_heap_empty(&psset->pageslabs[ret_ind])) {
+		bitmap_set(psset->bitmap, &psset_bitmap_info, ret_ind);
+	}
+	return ret;
+}
+
+static void
+psset_insert(psset_t *psset, edata_t *ps, size_t largest_range) {
+	pszind_t pind = sz_psz2ind(sz_psz_quantize_floor(
+	    largest_range << LG_PAGE));
+
+	assert(pind < PSSET_NPSIZES);
+
+	if (edata_heap_empty(&psset->pageslabs[pind])) {
+		bitmap_unset(psset->bitmap, &psset_bitmap_info, (size_t)pind);
+	}
+	edata_heap_insert(&psset->pageslabs[pind], ps);
+}
+
+/*
+ * Given a pageslab ps and an edata to allocate size bytes from, initializes the
+ * edata with a range in the pageslab, and puts ps back in the set.
+ */
+static void
+psset_ps_alloc_insert(psset_t *psset, edata_t *ps, edata_t *r_edata,
+    size_t size) {
+	size_t start = 0;
+	/*
+	 * These are dead stores, but the compiler will issue warnings on them
+	 * since it can't tell statically that found is always true below.
+	 */
+	size_t begin = 0;
+	size_t len = 0;
+
+	fb_group_t *ps_fb = edata_slab_data_get(ps)->bitmap;
+
+	size_t npages = size >> LG_PAGE;
+	size_t ps_npages = edata_size_get(ps) >> LG_PAGE;
+
+	size_t largest_unchosen_range = 0;
+	while (true) {
+		bool found = fb_urange_iter(ps_fb, ps_npages, start, &begin,
+		    &len);
+		/*
+		 * A precondition to this function is that ps must be able to
+		 * serve the allocation.
+		 */
+		assert(found);
+		if (len >= npages) {
+			break;
+		}
+		if (len > largest_unchosen_range) {
+			largest_unchosen_range = len;
+		}
+		start = begin + len;
+	}
+	uintptr_t addr = (uintptr_t)edata_base_get(ps) + begin * PAGE;
+	edata_init(r_edata, edata_arena_ind_get(r_edata), (void *)addr, size,
+	    /* slab */ false, SC_NSIZES, /* sn */ 0, extent_state_active,
+	    /* zeroed */ false, /* committed */ true, EXTENT_PAI_HPA,
+	    EXTENT_NOT_HEAD);
+	edata_ps_set(r_edata, ps);
+	fb_set_range(ps_fb, ps_npages, begin, npages);
+	/*
+	 * OK, we've got to put the pageslab back.  First we have to figure out
+	 * where, though; we've only checked run sizes before the pageslab we
+	 * picked.  We also need to look for ones after the one we picked.  Note
+	 * that we want begin + npages as the start position, not begin + len;
+	 * we might not have used the whole range.
+	 */
+	start = begin + npages;
+	while (start < ps_npages) {
```

Comment:
I think we can, although practically we don't lose a lot of time here in some tests. Mind leaving it as a TODO for now?

(I'm not ardently opposed or anything, it's just a little bit of a hassle since now callers have to track it).

Reply:
Sounds good. Agree that the first version should be more straightforward and leave the optimizations in future PRs.